File: frontend/src/lib/api/monitoring.ts
/**
 * Monitoring API: jobs, baselines, deviations, alerts, and stats.
 */

import { apiGet, apiPost } from "./client";

// -- Types --------------------------------------------------------------------

export interface MonitoringJobData {
  id: string;
  engagement_id: string;
  name: string;
  source_type: string;
  status: string;
  connection_id: string | null;
  baseline_id: string | null;
  schedule_cron: string;
  config: Record<string, unknown> | null;
  last_run_at: string | null;
  next_run_at: string | null;
  error_message: string | null;
}

export interface MonitoringJobList {
  items: MonitoringJobData[];
  total: number;
}

export interface BaselineData {
  id: string;
  engagement_id: string;
  process_model_id: string | null;
  name: string;
  element_count: number;
  process_hash: string | null;
  is_active: boolean;
  created_at: string;
}

export interface BaselineList {
  items: BaselineData[];
  total: number;
}

export interface DeviationData {
  id: string;
  engagement_id: string;
  monitoring_job_id: string;
  category: string;
  description: string;
  affected_element: string | null;
  magnitude: number;
  details: Record<string, unknown> | null;
  detected_at: string;
}

export interface DeviationList {
  items: DeviationData[];
  total: number;
}

export interface AlertData {
  id: string;
  engagement_id: string;
  monitoring_job_id: string;
  severity: string;
  status: string;
  title: string;
  description: string;
  deviation_ids: string[] | null;
  acknowledged_by: string | null;
  acknowledged_at: string | null;
  resolved_at: string | null;
  created_at: string;
}

export interface AlertList {
  items: AlertData[];
  total: number;
}

export interface MonitoringStats {
  active_jobs: number;
  total_deviations: number;
  open_alerts: number;
  critical_alerts: number;
}

// -- API functions ------------------------------------------------------------

export async function fetchMonitoringJobs(
  engagementId?: string,
): Promise<MonitoringJobList> {
  const params = engagementId ? `?engagement_id=${engagementId}` : "";
  return apiGet<MonitoringJobList>(`/api/v1/monitoring/jobs${params}`);
}

export async function fetchMonitoringJob(
  jobId: string,
): Promise<MonitoringJobData> {
  return apiGet<MonitoringJobData>(`/api/v1/monitoring/jobs/${jobId}`);
}

export async function createMonitoringJob(
  body: Record<string, unknown>,
): Promise<MonitoringJobData> {
  return apiPost<MonitoringJobData>("/api/v1/monitoring/jobs", body);
}

export async function fetchMonitoringStats(
  engagementId: string,
): Promise<MonitoringStats> {
  return apiGet<MonitoringStats>(
    `/api/v1/monitoring/stats/${engagementId}`,
  );
}

export async function fetchAlerts(
  engagementId?: string,
): Promise<AlertList> {
  const params = engagementId ? `?engagement_id=${engagementId}` : "";
  return apiGet<AlertList>(`/api/v1/monitoring/alerts${params}`);
}

export async function fetchDeviations(
  engagementId?: string,
): Promise<DeviationList> {
  const params = engagementId ? `?engagement_id=${engagementId}` : "";
  return apiGet<DeviationList>(`/api/v1/monitoring/deviations${params}`);
}

export async function fetchBaselines(
  engagementId?: string,
): Promise<BaselineList> {
  const params = engagementId ? `?engagement_id=${engagementId}` : "";
  return apiGet<BaselineList>(`/api/v1/monitoring/baselines${params}`);
}

File: src/api/schemas/monitoring.py
"""Pydantic schemas for monitoring API routes."""

from __future__ import annotations

from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from src.core.models import MonitoringSourceType, MonitoringStatus

# ---------------------------------------------------------------------------
# Monitoring Job Schemas
# ---------------------------------------------------------------------------


class MonitoringJobCreate(BaseModel):
    engagement_id: UUID
    name: str = Field(..., min_length=1, max_length=255)
    source_type: MonitoringSourceType
    connection_id: UUID | None = None
    baseline_id: UUID | None = None
    schedule_cron: str = "0 0 * * *"
    config: dict[str, Any] | None = None


class MonitoringJobUpdate(BaseModel):
    name: str | None = None
    schedule_cron: str | None = None
    config: dict[str, Any] | None = None
    status: MonitoringStatus | None = None


class MonitoringJobResponse(BaseModel):
    id: str
    engagement_id: str
    name: str
    source_type: str
    status: str
    connection_id: str | None = None
    baseline_id: str | None = None
    schedule_cron: str
    config: dict[str, Any] | None = None
    last_run_at: str | None = None
    next_run_at: str | None = None
    error_message: str | None = None


class MonitoringJobList(BaseModel):
    items: list[MonitoringJobResponse]
    total: int


# ---------------------------------------------------------------------------
# Baseline Schemas
# ---------------------------------------------------------------------------


class BaselineCreate(BaseModel):
    engagement_id: UUID
    process_model_id: UUID | None = None
    name: str = Field(..., min_length=1, max_length=255)
    snapshot_data: dict[str, Any] | None = None


class BaselineResponse(BaseModel):
    id: str
    engagement_id: str
    process_model_id: str | None = None
    name: str
    element_count: int
    process_hash: str | None = None
    is_active: bool
    created_at: str


class BaselineList(BaseModel):
    items: list[BaselineResponse]
    total: int


# ---------------------------------------------------------------------------
# Deviation Schemas
# ---------------------------------------------------------------------------


class DeviationResponse(BaseModel):
    id: str
    engagement_id: str
    monitoring_job_id: str
    category: str
    description: str
    affected_element: str | None = None
    magnitude: float
    details: dict[str, Any] | None = None
    detected_at: str


class DeviationList(BaseModel):
    items: list[DeviationResponse]
    total: int


# ---------------------------------------------------------------------------
# Alert Schemas
# ---------------------------------------------------------------------------


class AlertResponse(BaseModel):
    id: str
    engagement_id: str
    monitoring_job_id: str
    severity: str
    status: str
    title: str
    description: str
    deviation_ids: list[str] | None = None
    acknowledged_by: str | None = None
    acknowledged_at: str | None = None
    resolved_at: str | None = None
    created_at: str


class AlertList(BaseModel):
    items: list[AlertResponse]
    total: int


class AlertActionRequest(BaseModel):
    action: str = Field(..., pattern="^(acknowledge|resolve|dismiss)$")
    actor: str = "system"


# ---------------------------------------------------------------------------
# Stats Schema
# ---------------------------------------------------------------------------


class MonitoringStats(BaseModel):
    active_jobs: int
    total_deviations: int
    open_alerts: int
    critical_alerts: int


# ---------------------------------------------------------------------------
# Dashboard Schemas (Story #371)
# ---------------------------------------------------------------------------


class AgentStatusResponse(BaseModel):
    """Agent health status summary."""

    total: int
    healthy: int
    degraded: int
    unhealthy: int
    agents: list[dict[str, Any]]


class DeviationCountResponse(BaseModel):
    """Deviation counts by severity."""

    total: int
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0
    info: int = 0


class AlertSummaryResponse(BaseModel):
    """Alert summary for dashboard."""

    total_open: int
    new: int
    acknowledged: int
    critical_open: int


class ComplianceDataPointResponse(BaseModel):
    """Single compliance score data point."""

    date: str
    score: float


class ComplianceTrendResponse(BaseModel):
    """Compliance score trend for dashboard."""

    current_score: float
    trend_direction: str
    data_points: list[ComplianceDataPointResponse]


class DashboardResponse(BaseModel):
    """Complete monitoring dashboard aggregated data."""

    engagement_id: str
    date_from: str
    date_to: str
    agent_status: AgentStatusResponse
    deviations: DeviationCountResponse
    evidence_flow_rate: float
    alerts: AlertSummaryResponse
    compliance_trend: ComplianceTrendResponse


# ---------------------------------------------------------------------------
# Pipeline Metrics Schema (Story #360)
# ---------------------------------------------------------------------------


class PipelineMetricsResponse(BaseModel):
    """Response from pipeline metrics endpoint."""

    processing_rate: float
    queue_depth: int
    p99_latency_ms: float
    avg_latency_ms: float
    total_processed: int
    total_errors: int
    avg_quality: float
    window_seconds: int


# ---------------------------------------------------------------------------
# Agent Health Schemas (Story #346)
# ---------------------------------------------------------------------------


class AgentHealthEntry(BaseModel):
    """Health status for a single monitoring agent."""

    agent_id: str
    status: str
    last_poll_time: str | None = None
    items_processed_total: int = 0
    consecutive_failures: int = 0
    watermark: str | None = None


class AgentHealthResponse(BaseModel):
    """Response for agent health endpoint."""

    agents: list[AgentHealthEntry]
    total: int
    healthy: int
    degraded: int
    unhealthy: int
